docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+# Bridging generic constraints
+
+Since introduction of generics in .NET there hasn't been a way to go from less contrained type parameter T to a more constrained type parameter U. If U has any constraints that T doesn't have, a T cannot be substituted for U. We've seen many instances where it would be useful to allow this for concrete substitution of T that match U's constraints. There's no way to do this substitution at runtime besides using reflection right now.
+
+One way to go about this would be to relax constraint checks within method bodies so that constraints are validated at runtime instead of at compile time:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+static void DoConstrained<T>() where T : struct { }
+
+static void Do<T>
+{
+    if (typeof(T).IsValueType)
+    {
+        // The T in Do is unconstrainted, but DoConstrained requires it to be a struct
+        // The run time check above ensures this.
+        // This will not compile as C# today because the constraints don't match.
+        // Today, it will also result in a verification exception at runtime when compiling the Do method.
+        DoConstrained<T>();
+    }
+}
+```
+
+The purpose of this document is to sketch out a possible implementation on the runtime side that would allow above code to work.
+
+This feature doesn't assume any kind of relaxation of constraint checks outside method bodies. E.g. allowing to derive from `class Base<T> where T : IFoo { }` with `class Derived<T> : Base<T> { }` to load for `Derived<IFoo>` is not in scope.
```

Comment:
I had a proposal for something like this, just FYI: https://github.com/dotnet/csharplang/discussions/1992

Reply:
> er, there's something odd about the above. T is not an instance, its a type, but you are putting an instance into the a var

Fixed.

Reply:
> One of the common problems is a need to match and decompose generic type:

I'll add that as out of scope. Would need to rename this proposal because that's no longer "bridging constraints".

I don't see a way how we could encode that without actually extending the file format in a structurally-compatible way. (So far our extensions to the file format didn't require changes to e.g. signature parsing, but this might).

Reply:
It may be noteworthy that this feature would create an interesting issue that would need to be handled somehow:
```cs
class C : IEnumerable<int>, IEnumerable<string> { }

void M<T>(T o)
{
    if (T is IEnumerable<U>) // what do we do here?
    {
    }
}
```